server/plugin.go

```diff
@@ -27,11 +32,42 @@ type Plugin struct {
 	// setConfiguration for usage.
 	configuration *configuration
 
+	// activatedLock synchronizes access to activated.
```

Comment:
Having three locks "smells" wrong to me. Why not just one? `connectedLock ` is going to be in contention by potentially lots of users, while `surveyLock` seems to happen on just configuration changes, so it's not clear we gain much by having so much locking. (Commented separately above, but I also think `activatedLock ` is redundant.)

Reply:
`connectedLock` was an initial attempt at the logic that the random delay solves. I'm aware that it has contention issues, but I couldn't think of another way to do that where each user has their own lock without allocating a lock for each user. I figured it would be okay for the initial version of this until we have some better per-user synchronization logic such as the example I posted [in the developer toolkit channel](https://community-daily.mattermost.com/core/pl/64ipipkd7f8zxm3h9ez9u3f8da).

`surveyLock` is needed because `checkForNextSurvey` has to run in a goroutine to avoid `OnActivate` timing out, so it's possible that it gets called twice if the configuration changes after activating the plugin. I'm also not sure if we ensure that `OnConfigurationChanged` is called synchronously, so I think it might be needed there as well.